### DISCUSSION TITLE
Non-deterministic results for minimum_weight_full_matching

### DISCUSSION THREAD

**Discussion by Adamits** [networkx/networkx#4634] [Q&A]
Hi,

When I use networkx.algorithms.bipartite.matching.minimum_weight_full_matching on a graph that presumably has multiple alignments that both lead to an optimal solution, I get different alignments for different runs. That is, If I run minimum_weight_full_matching on the same bipartite graph (whose 2 sets have unequal cardinality), I can get different matchings in 2 duplicate experiments, they seem to sometimes yield different matchings.

Is this expected? What is the source of the randomness?

Thanks for any tips!

**Comment by Adamits** [comment]
Update:

1. I add edges to make the 2 sets of the bipartite graph fully connected
2. Many edges have a weight of 0
3. When I replace the networkx code, by building the rectangular cost matrix myself, and calling `scipy.optimize.linear_sum_assignment` myself, I no longer get the varying results (which is what I expect to happen).

I am thinking the culprit could be the line in https://networkx.org/documentation/stable/_modules/networkx/algorithms/bipartite/matching.html#minimum_weight_full_matching: 
```weights = np.full(weights_sparse.shape, np.inf)```
Does this create a square matrix, where nonexistent edges get infinity weights? Could this cause some variation in results?

**Comment by dschult** [comment]
The use of `np.inf` is not the source of variation in results. The variation is almost certainly due to variation in the ordering of the sets. Sets in Python are not ordered. The order depends on a hash function. I believe that hashes of strings are generated based on a random number created at the beginning of the python session. So, from one python session to another, set order may differ, but within one session it should be the same. [Edit: you can make it consistent from session to session using the PYTHONHASHSEED environment variable.]

Does any of this seem to apply in your case?

We could always use dicts to mimic ordered sets. But sets are used throughout the codebase so I'd like to understand the problem before jumping to that kind of solution.

**Reply by Adamits** [reply]
Hi, thanks for the response!

I was adding nodes to the graph with `graph.add_nodes_from(nodes)` where `nodes`  are a frozenset of strings, so the hashing of strings definitely sounds like it could be the culprit!

How could I confirm this? Should I try by using objects with a different hashing function as nodes?

**Reply by dschult** [reply]
Try setting the environmental variable PYTHONHASHSEED to 0, start the interpretter and run your script.
Then try again with PYTHONHASHSEED=1 and run again. Repeat until it changes the output. Rerun with the
different values to convince yourself that the same seed value gives the same result.

In bash:  
```
export PYTHONHASHSEED=0
python myscript.py 
export PYTHONHASHSEED=1
python myscript.py 
...
```
If not in bash, google pythonhashseed to find out how to set it.
